staging/src/k8s.io/client-go/dynamic/client.go

```diff
@@ -246,9 +282,11 @@ func (rc *ResourceClient) Watch(opts metav1.ListOptions) (watch.Interface, error
 
 func (rc *ResourceClient) Patch(name string, pt types.PatchType, data []byte) (*unstructured.Unstructured, error) {
 	result := new(unstructured.Unstructured)
+	resourceName, subresourceName := rc.parseResourceSubresourceName()
 	err := rc.cl.Patch(pt).
 		NamespaceIfScoped(rc.ns, rc.resource.Namespaced).
-		Resource(rc.resource.Name).
+		Resource(resourceName).
+		SubResource(subresourceName...).
```

Comment:
Would be better to have `subresources` as an argument...that's how we do in typed clients.

https://github.com/kubernetes/kubernetes/blob/ae0f45e855381d3731f8e66e536d1ad093ea9948/pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/statefulset.go#L164-L176

Reply:
I agree that we should emulate typed clients behavior in dynamic client.

However, I think `rc.resource.Name` (which is a `metav1.APIResource.Name`) should allow names containing '/', e.g. `pods/binding`, to reflect what we are serving in api discovery: 
    {
      "name": "pods/binding",
      "singularName": "",
      "namespaced": true,
      "kind": "Binding",
      "verbs": [
        "create"
      ]
    },

In that case, should we parse `rc.resource.Name` and change `Resource(rc.resource.Name).` into `Resource(resourceName).`?

Reply:
@roycaihw can you try what happens if user creates a clientpool, and then call [clientpool.ClientForGroupVersionResource()](https://github.com/kubernetes/client-go/blob/master/dynamic/client_pool.go#L82) with a subresource? Can the restmapper handle subresource?

Reply:
Yes, the restmapper can handle subresource: https://github.com/kubernetes/kubernetes/pull/57171/commits/9f9ec171152b2b5ce0ab2da6ad90bba4935627e2#diff-b9e0984a44b8dbba37804e9a762ea6cbR89. It seems to return a client for GV https://github.com/kubernetes/kubernetes/blob/92e4d3da0076f923a45d54d69c84e91ac6a61a55/staging/src/k8s.io/client-go/dynamic/client_pool.go#L86 only, if there is no matching resource.

Reply:
@nikhita a common use case is constructing the dynamic client with the result returned by the discovery client. In this case, it's convenient if the [client.Resource(resource, namespace)](https://github.com/kubernetes/client-go/blob/master/dynamic/client.go#L122) allows `resource` to be in the form of `resource/subresource`. Typed clients don't have this use case so I think we can tolerate the discrepancy.

Reply:
:+1: